tox.ini

```diff
@@ -83,39 +76,10 @@ deps =
     # test-requirements
     httpretty==0.9.6
     mock==1.3.0
-    nose==1.3.7
+    pytest==2.8.7
```

Comment:
Xenial is busted for me: I think we need 3.0.7

Reply:
We don't need or want 3.0.7, because that's not the version in xenial:

```shell
$ rmadison python-pytest
...
 python-pytest | 2.8.7-4        | xenial/universe         | all
...
```

The problem here is that, not especially surprisingly, the version of pytest in xenial doesn't work with a more recent version of Python. pytest 2.8.7 does work fine on xenial's version of Python, because we have this tox target in our Travis configuration, and the build is green for this PR: https://travis-ci.org/canonical/cloud-init/jobs/652102211

I think the "fix" here is to accept that we can no longer run the xenial tox env on focal. (This case has always been of limited utility because unless you're running xenial, you aren't actually testing a configuration that we ever expect to be deployed; no-one is running xenial's libraries with Python 3.8 in the real world.) As such, I don't think this is a great loss. I'll push up a change to the default tox targets that we run.

Does that sound OK?

Reply:
Your argument above I think boils down to "Distro x has a different version of python than the developer's system.   Therefore test of distro x in tox is of no use.".    Taken to its limits, the argument basically defeats the entire intent of tox.  Tox does not reproduce exactly the target of *any* environment, but usefully and easily approximates them.

The xenial tox environment was never a *perfect* environment.  But it caught a very large number of errors that would only otherwise be exposed in much heavier test environment (a container or package build).

The value provided was less about testing a specific *python* but more about testing dependencies at the levels specified.  Ie we have restricted our use of `requests` to version 2.9.1.  If you used some feature of requests not present in that version, then test would fail. Tox did a good job of quickly finding failure where we used feature or behavior of a library at one version that was not available at a different version.

I don't have a good answer.   But just dropping the environment will result in a higher number of developer failures being found by c-i that previously were found on developers run of 'tox'.

Reply:
OK, I've just pushed up https://github.com/canonical/cloud-init/commit/8affaf250686b3264dab68745fbf7c23e154fe1a which addresses this much better, I think. (For some reason, the PR hasn't updated, GitHub is having some Issues.)